buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java

```diff
@@ -237,22 +237,29 @@ private boolean allocate(int size, int maxCapacity, Thread currentThread, Adapti
             int index = (int) (threadId & mask);
             for (int i = 0, m = Integer.numberOfTrailingZeros(~mask); i < m; i++) {
                 Magazine mag = mags[index + i & mask];
+                if (buf == null) {
+                    grabbedBuffer = true;
+                    buf = mag.newBuffer();
+                }
                 if (mag.tryAllocate(size, sizeBucket, maxCapacity, buf)) {
                     // Was able to allocate.
-                    return true;
+                    return buf;
                 }
             }
             expansions++;
         } while (expansions <= EXPANSION_ATTEMPTS && tryExpandMagazines(mags.length));
-        return false;
+        if (grabbedBuffer) {
+            buf.release();
+        }
+        return null;
     }
 
     /**
      * Allocate into the given buffer. Used by {@link AdaptiveByteBuf#capacity(int)}.
      */
     void allocate(int size, int maxCapacity, AdaptiveByteBuf into) {
         Magazine magazine = into.chunk.magazine;
-        if (!allocate(size, maxCapacity, Thread.currentThread(), into)) {
+        if (allocate(size, maxCapacity, Thread.currentThread(), into) == null) {
```

Comment:
imho this usage is really odd as if this methods returns non null its a `ByteBuf` which has a lifetime. I know that this would just be the same as `into` and so its fine to not release but this is more an implementation detail and very confusing and fragile.